log/log.go

```diff
@@ -14,34 +16,98 @@ import (
 var (
 	log      *zap.SugaredLogger
 	errorLog *os.File
+	output   string
+	logLevel sync.Map // this is a thread-safe map[string]string
+)
+
+const (
+	defaultLevel  = "*:error"
+	defaultOutput = "stderr"
 )
 
 func init() {
 	// Allow overriding the default log level via $LOG_LEVEL, so that the
 	// environment variable can be set globally even when running tests.
 	// Always initializing the logger is also useful to avoid panics when
 	// logging if the logger is nil.
-	level := "error"
+	level := []string{defaultLevel}
 	if s := os.Getenv("LOG_LEVEL"); s != "" {
-		level = s
+		level = strings.Split(s, ",")
 	}
-	Init(level, "stderr")
+
+	InitWithLevels(level, defaultOutput)
+}
+
+// Init initializes the logger, with a single logLevel (debug|info|warn|error|fatal).
+// out can be either "stdout", "stderr" or "/path/to/file" and it's a global setting.
+//
+// kept for backwards compatibility, it just wraps InitWithLevels()
+func Init(logLevel string, out string) {
+	InitWithLevels([]string{logLevel}, out)
 }
 
-func Logger() *zap.SugaredLogger { return log }
+// InitWithLevels initializes the logger.
+// logLevels are strings like "somename:debug" or "*:info" (equivalent to simply "info").
+// out can be either "stdout", "stderr" or "/path/to/file" and it's a global setting.
+//
+// It's safe to use concurrently.
+func InitWithLevels(logLevels []string, out string) {
+	for _, ll := range logLevels {
+		split := strings.Split(ll, ":")
+		if len(split) > 1 {
+			// split[0] is subsystem name, split[1] is level
+			logLevel.Store(split[0], split[1])
+		} else {
+			// backwards compatibility: --logLevel=error is equivalent to --logLevel="*:error"
+			logLevel.Store("*", split[0])
+		}
+	}
+
+	output = out
+	log = newLogger(LogLevelFor("*")).Desugar().WithOptions(zap.AddCallerSkip(1)).Sugar()
 
-// Init initializes the logger. Output can be either "stdout/stderr/filePath"
-func Init(logLevel string, output string) {
-	cfg := newConfig(logLevel, output)
+	log.Infof("logger construction succeeded at default level %s with output %s", LogLevelFor("*"), output)
+}
 
+// LogLevelFor returns the configured log level for the named subsystem,
+// or "" if the subsystem has no specific log level configured.
+//
+// It's safe to use concurrently.
+func LogLevelFor(name string) string {
+	value, _ := logLevel.Load(name)
+	if value, ok := value.(string); ok {
+		return value
+	}
+	return ""
+}
+
+// Named returns a named logger
+func Named(name string) *zap.SugaredLogger {
+	level := LogLevelFor(name)
+	if level == "" {
+		level = LogLevelFor("*")
+	}
+	defer log.Infof("logger construction succeeded for subsystem %s at level %s", name, level)
+	return newLogger(level).Named(name)
+}
+
+// Logger returns the unwrapped logger
+func Logger() *zap.SugaredLogger {
```

Comment:
How about calling this `Unnamed` now, so then we end up with `log.Named(...)` and `log.Unnamed()`?

Reply:
dropped this func